base/staticint.jl

```diff
@@ -0,0 +1,105 @@
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Nothing} = promote_type(T1, Int)
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Missing} = promote_type(T1, Int)
+for T in [:Bool, :Missing, :BigFloat, :BigInt, :Nothing, :Any]
+    # let S = :Any
+    @eval begin
+        Base.promote_rule(@nospecialize(S::Type{<:StaticInt}), ::Type{$T}) = promote_type(Int, $T)
+        Base.promote_rule(::Type{$T}, @nospecialize(S::Type{<:StaticInt})) = promote_type($T, Int)
+    end
+end
+Base.promote_rule(@nospecialize(T1::Type{<:StaticInt}), @nospecialize(T2::Type{<:StaticInt})) = Int
+
+Base.:(%)(@nospecialize(n::StaticInt), ::Type{Integer}) = Int(n)
+
+Base.eltype(@nospecialize(T::Type{<:StaticInt})) = Int
+Base.iszero(::Zero) = true
+Base.iszero(@nospecialize(x::StaticInt)) = false
+Base.isone(::One) = true
+Base.isone(@nospecialize(x::StaticInt)) = false
+Base.zero(@nospecialize(x::Type{<:StaticInt})) = Zero()
+Base.one(@nospecialize(x::Type{<:StaticInt})) = One()
+
+for T in [:Real, :Rational, :Integer]
+    @eval begin
+        @inline Base.:(+)(x::$T, @nospecialize(y::StaticInt)) = x + Int(y)
+        @inline Base.:(+)(@nospecialize(x::StaticInt), y::$T) = Int(x) + y
+        @inline Base.:(-)(x::$T, @nospecialize(y::StaticInt)) = x - Int(y)
+        @inline Base.:(-)(@nospecialize(x::StaticInt), - y::$T) = Int(x) - y
+        @inline Base.:(*)(x::$T, @nospecialize(y::StaticInt)) = x * Int(y)
+        @inline Base.:(*)(@nospecialize(x::StaticInt), y::$T) = Int(x) * y
+    end
+end
+@inline Base.:(-)(::StaticInt{M}) where {M} = StaticInt{-M}()
+
+for f in [:(+), :(-), :(*), :(÷), :(%), :(<<), :(>>), :(>>>), :(&), :(|), :(⊻)]
+    @eval begin
+        @inline Base.$f(::StaticInt{M}, ::StaticInt{N}) where {M,N} = StaticInt{$f(M,N)}()
+    end
+end
+for f in [:(<<), :(>>), :(>>>)]
+    @eval begin
+        @inline Base.$f(@nospecialize(x::StaticInt), y::UInt) where {M} = $f(Int(x), y)
+        @inline Base.$f(x::Integer, @nospecialize(y::StaticInt)) = $f(x, Int(y))
+    end
+end
+for f in [:(==), :(!=), :(<), :(≤), :(>), :(≥)]
+    @eval begin
+        @inline Base.$f(::StaticInt{M}, ::StaticInt{N}) where {M,N} = $f(M, N)
+        @inline Base.$f(@nospecialize(x::StaticInt), y::Int) where {M} = $f(Int(x), y)
+        @inline Base.$f(x::Int, @nospecialize(y::StaticInt)) = $f(x, Int(y))
+    end
+end
+
+@inline Base.widen(@nospecialize(x::StaticInt)) = Int(x)
+
+Base.UnitRange{T}(@nospecialize(start::StaticInt), stop) where {T<:Real} = UnitRange{T}(T(start), T(stop))
+Base.UnitRange{T}(start, @nospecialize(stop::StaticInt)) where {T<:Real} = UnitRange{T}(T(start), T(stop))
+function Base.UnitRange{T}(@nospecialize(start::StaticInt), @nospecialize(stop::StaticInt)) where {T<:Real}
+    UnitRange{T}(T(start), T(stop))
+end
+
+Base.UnitRange(@nospecialize(start::StaticInt), stop) = UnitRange(Int(start), stop)
```

Comment:
Should we make `UnitRange(One(), 10) === OneTo(10)`?

Reply:
My understanding has been that if we have a type call then we return that type by convention.